Add cypress e2e tests

nprbst opened the pull request:
@HipsterBrown ready for some feedback...

Commit:
WIP: Good headway...

Commit:
Refactored to a happy place.

Remaining work is just to complete the test states and events...

Commit:
Covered Editor with tests via nested machine.

Required some refactoring to keep things clean and modular.

HipsterBrown commented on src/utils/env.ts:
```diff
@@ -1 +1,4 @@
 export const isProd = () => process.env.NODE_ENV === "production";
+
+// @ts-expect-error
+export const isCypress = () => window && window.Cypress;
```
Optional chaining is useful here:

```suggestion
export const isCypress = () => window?.Cypress;
```

HipsterBrown commented on cypress/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "../tsconfig.json",
+  "include": ["./**/*.ts", "integration/sample.spec.js"],
```
```suggestion
  "include": ["./**/*.ts"],
```

HipsterBrown commented on yarn.lock:
```diff
@@ -0,0 +1,7441 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
This repo is using `npm` rather than `yarn`, so can you remove this file and run `npm install` to update the `package-lock.json`?

HipsterBrown commented on src/App.tsx:
```diff
@@ -15,9 +15,10 @@ import { Settings } from "./pages/Settings";
 
-if (!isProd()) {
+// Do not start the inspector in Production or within Cypress
+if (!isProd() && !isCypress()) {
```
:+1:

HipsterBrown reviewed (commented):
This is a great start! The interceptors are really neat, something I didn't know about for Cypress.

One of the main things that stands out is the abstractions under the `xstateTestUtils`, that make it kind of hard to understand the setup for testing model and how it applies to the subject under test. 

It would be helpful for me, and probably other folks learning from this repo in the future, to see the manual setup of each machine to the test model. For instance, I wouldn't know where to start to contribute another feature spec.

Commit:
Covered Settings and Comments with tests via nested machines.

Further refactoring to support KitchenSink specs composed from nested-machine specs.

Commit:
Address PR feedback.

nprbst commented:
@HipsterBrown I hear what you're saying about the additional abstractions in `xstateTestUtils`...however, they are what allow the test plans to be _composed_ by stitching together machines, events and tests. I'd like to take a crack at addressing this hurdle via a `cypress/README.md`, and possibly some additional refactoring. But, I've kinda blown my time allotment for this effort...need to pause for a few days before finishing this PR.

HipsterBrown commented:
> I'd like to take a crack at addressing this hurdle via a cypress/README.md

@nprbst That would also be very helpful! No rush on finishing this up, I appreciate any time you're able to put toward the contribution. 😄

nprbst commented:
Sorry I've been slow to pick this back up. Hopeful I'll have time this week!
